### PR TITLE
Event grid track 2

### PR DESCRIPTION
**Pull request by lmazuel**
Known issues:
- [x]  sdist validation (MANIFEST.in)
- [x]  offline preparer
- [x]  Py 2.7 urllib parse
- [x] encoding compatibility with Py2.7

### CODE REVIEW

File: sdk/eventgrid/azure-eventgrid/CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # Release History
 
+## 2.0.0 (2020-08-07)
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/CHANGELOG.md:3]:
```suggestion
## 2.0.0b1 (2020-08-07)
```

File: sdk/eventgrid/azure-eventgrid/CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,9 @@
 # Release History
 
+## 2.0.0b1 (2020-08-07)
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/CHANGELOG.md:3]:
```suggestion
## 2.0.0 (2020-08-07)
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    if not topic_hostname.startswith('https://'):
+        topic_hostname = "https://{}".format(topic_hostname)
+    if not topic_hostname.endswith("/api/events"):
+        topic_hostname = "{}/api/events".format(topic_hostname)
+    
+    return topic_hostname
+
+def _hmac(key, message):
+    decoded_key = base64.b64decode(key)
+    bytes_message = bytes(message, 'ascii')
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py:60]:
```suggestion
    bytes_message = message.encode('ascii')
```

File: sdk/eventgrid/azure-eventgrid/CHANGELOG.md (hunk 1 above)
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/CHANGELOG.md:3]:
```suggestion
## 2.0.0b1 (Unreleased)
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_version.py [hunk 4]
```diff
@@ -9,5 +9,4 @@
 # regenerated.
 # --------------------------------------------------------------------------
 
-VERSION = "1.3.0"
-
+VERSION = "2.0.0"
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_version.py:12]:
```suggestion
VERSION = "2.0.0b1"
```

File: sdk/eventgrid/azure-eventgrid/CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,9 @@
 # Release History
 
+## 2.0.0b1 (Unreleased)
+
+  - Placeholder - NEEDS TO BE CHANGED
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/CHANGELOG.md:5]: Were you planning to swap this out with some representative snippet?  or do it as an intentional step later once we're more settled.

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_consumer.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        """Single event following CloudEvent/EventGridEvent schema will be parsed and returned as DeserializedEvent.
+        :param event: The event to be deserialized.
+        :type event: Union[str, dict, bytes]
+        :rtype: models.DeserializedEvent
+
+        :raise: :class:`ValueError`, when events do not follow CloudEvent or EventGridEvent schema.
+        """
+        try:
+            if isinstance(event, six.binary_type):
+                event = json.loads(event.decode('utf-8'))
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_consumer.py:41]: Should we allow encoding to be provided?

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    if topic_hostname.startswith('http://'):
+        raise ValueError("HTTP is not supported. Only HTTPS is supported.")
+    if not topic_hostname.startswith('https://'):
+        topic_hostname = "https://{}".format(topic_hostname)
+    if not topic_hostname.endswith("/api/events"):
+        topic_hostname = "{}/api/events".format(topic_hostname)
+    
+    return topic_hostname
+
+def _hmac(key, message):
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py:58]: nit: functions should have imperitive names; so "generate_hmac"?

Comment (KieranBrantnerMagee) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py:58]: quick 2nd note on this:  There's a push to have type signatures for internal functions.  I'm not gonna fine-toothed-comb on this now since we're in a very early preview, but my learnings from SB is that life is a lot easier of one just does the typing/docstring bits as one goes along.

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_models.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        'type': {'key': 'type', 'type': 'str'},
+        'time': {'key': 'time', 'type': 'iso-8601'},
+        'specversion': {'key': 'specversion', 'type': 'str'},
+        'dataschema': {'key': 'dataschema', 'type': 'str'},
+        'datacontenttype': {'key': 'datacontenttype', 'type': 'str'},
+        'subject': {'key': 'subject', 'type': 'str'},
+    }
+
+    def __init__(self, source, type, **kwargs):
+        # type: (Any) -> None
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_models.py:63]: this type hint seems incorrect

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_models.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+class DeserializedEvent():
+    """The container for the deserialized event model and mapping of event envelope properties.
+        :param dict event: dict
+    """
+    # class variable
+    _event_type_mappings = {'Microsoft.Storage.BlobCreated': StorageBlobCreatedEventData}
+
+    def __init__(self, event):
+        # type: (Any) -> None
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_models.py:140]: as with other __init__ hints.

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        auth_policy = _get_authentication_policy(credential)
+        self._client = EventGridPublisherClientImpl(authentication_policy=auth_policy, **kwargs)
+
+    def send(self, events, **kwargs):
+        # type: (Union[List[CloudEvent], List[EventGridEvent], List[CustomEvent]], Any) -> None
+        """Sends event data to topic hostname specified during client initialization.
+
+        :param events: A list of CloudEvent/EventGridEvent/CustomEvent to be sent.
+        :type events: Union[List[models.CloudEvent], List[models.EventGridEvent], List[models.CustomEvent]]
+        :rtype: None
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:46]: Some places we have :rtype: None, other types we emit rtype when we don't return.  I don't recall any specific guidance here, but it seems to beg that question.
cc @lmazuel

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+         """
+
+        if all(isinstance(e, CloudEvent) for e in events):
+            self._client.publish_cloud_event_events(self._topic_hostname, events, **kwargs)
+        elif all(isinstance(e, EventGridEvent) for e in events):
+            self._client.publish_events(self._topic_hostname, events, **kwargs)
+        elif all(isinstance(e, CustomEvent) for e in events):
+            serialized_events = [dict(e) for e in events]
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:54]: It's a little interesting that we'd do this dictification here and not within the publish_custom_events flow (to make that function more "Accepting", which is 'broadly pythonic')

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_shared/mixins.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+# pylint:disable=protected-access
+
+class DictMixin(object):
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_shared/mixins.py:7]: @lmazuel I wish we had like, "python_sdk_utils" that contained generic helpers like this, since I'm 100% sure this exists in multiple of our libs.

(I realize this statement sounds like airy musings and not something actionable, so let me rephrase: "is this an insane proposition to you?")

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+        if all(isinstance(e, CloudEvent) for e in events):
+            await self._client.publish_cloud_event_events(self._topic_hostname, events, **kwargs)
+        elif all(isinstance(e, EventGridEvent) for e in events):
+            await self._client.publish_events(self._topic_hostname, events, **kwargs)
+        elif all(isinstance(e, CustomEvent) for e in events):
+            serialized_events = [dict(e) for e in events]
+            await self._client.publish_custom_event_events(self._topic_hostname, serialized_events, **kwargs)
+        else:
+            raise Exception("Event schema is not correct. Please send as list of all CloudEvents, list of all EventGridEvents, or list of all CustomEvents.")
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:54]: Probably should be a TypeError or ValueError depending on how we look at this. (I might lean towards TypeError, but this one could go either way, feel free to form your own opinion based on the python built-ins expectations and precedent.)

File: sdk/eventgrid/azure-eventgrid/samples/champion_scenarios/cs1_publish_custom_events_to_a_topic.py [hunk 14]
```diff
@@ -0,0 +1,19 @@
+from azure.eventgrid import EventGridPublisherClient, EventGridEvent, CloudEvent
+from azure.core.credentials import AzureKeyCredential
+
+topic_hostname = "<YOUR-TOPIC-NAME>.<REGION-NAME>-1.eventgrid.azure.net"
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/samples/champion_scenarios/cs1_publish_custom_events_to_a_topic.py:4]: Typically pluggables are os.environ loaded.  This is low-pri however, use your judgement for if you want to change these now, or if you feel it'll change a lot, add an issue to the backlog to make sure this is dealt with "sooner or later"

File: sdk/eventgrid/azure-eventgrid/samples/consume_samples/e2e_samples/consume_cloud_events_from_service_bus_queue.py [hunk 15]
```diff
@@ -0,0 +1,37 @@
+import sys
+import os
+
+PACKAGE_PARENT = '..'
+SCRIPT_DIR = os.path.dirname(os.path.realpath(os.path.join(os.getcwd(), os.path.expanduser(__file__))))
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/samples/consume_samples/e2e_samples/consume_cloud_events_from_service_bus_queue.py:5]: So I think I know why this is here, and _it's kinda a neat idea_ since running these samples ad-hoc (given the fact that they use real-package-paths from a subdir) often requires making sure you're in a proper env/path to actually resolve these imports can sometimes be a stumbling block.  however, I worry that if we start linking these samples to users, they'll be like "why would I have that."

Assuming these samples can be run as "self-contained-one-off-scripts" within a properly configured environment without this path, I'd probably yank this; although I feel bad saying this since it is a nice dev convenience layer.  Let me know if you have any clever ways to have our cake and eat it too, or if I'm missing something obvious.

File: sdk/eventgrid/azure-eventgrid/samples/consume_samples/e2e_samples/functionsapp/.funcignore [hunk 16]
```diff
@@ -0,0 +1,5 @@
+.git*
+.vscode
+local.settings.json
+test
+.venv
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/samples/consume_samples/e2e_samples/functionsapp/.funcignore:5]: do we want this to be the .funcignore everyone has? it seems a tiny bit special cased (ala .venv/test, but that may be intentional)

File: sdk/eventgrid/azure-eventgrid/samples/consume_samples/e2e_samples/functionsapp/.gitignore [hunk 17]
```diff
@@ -0,0 +1,130 @@
+# Byte-compiled / optimized / DLL files
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/samples/consume_samples/e2e_samples/functionsapp/.gitignore:1]: this is a massive gitignore.  maybe just a copy of our core one? which would make me question why we need it at the leaf here.  as with .funcignore wondering if we can be targeted if these are required; but also as with funcignore, this is very low pri so if it becomes a rathole feel free to punt on it.

File: sdk/eventgrid/azure-eventgrid/samples/consume_samples/e2e_samples/functionsapp/EventGridTrigger1/__init__.py [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+        'data': event.get_json(),
+        'topic': event.topic,
+        'subject': event.subject,
+        'event_type': event.event_type
+    })
+    logging.info(result)
+    #consumer = EventGridConsumer()
+    #deserialized_event = consumer.deserialize_events(result)
+    ## can only be EventGridEvent
+    #print("model: {}".format(event.model))
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/samples/consume_samples/e2e_samples/functionsapp/EventGridTrigger1/__init__.py:22]: deprecated?

File: sdk/eventgrid/azure-eventgrid/swagger/README.PYTHON_T2.md [hunk 19]
```diff
@@ -0,0 +1,21 @@
+# Azure EventGrid Client Generation for Python
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/swagger/README.PYTHON_T2.md:1]: I'm paranoid since Mayuri was not pleased when the SB swagger founds its way into Master; @lmazuel Is having this generation readme here kosher to your eyes?

File: sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+                                                     playback_fake_resource=playback_fake_resource,
+                                                     client_kwargs=client_kwargs)
+        self.resource_group_parameter_name = resource_group_parameter_name
+        self.parameter_name = parameter_name
+        self.parameter_location = parameter_location
+        self.name_prefix = name_prefix
+        if random_name_enabled:
+            self.resource_moniker = self.name_prefix + "egtopic"
+        
+        self.set_cache(use_cache, name_prefix)
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py:50]: name_prefix isn't necessary for caching since it doesn't change the nature of the resource.  (SKU for instance, or region, would change the resource)

File: sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+                name,
+                topic,
+                {}
+            )
+            self.resource = topic_operation.result()
+            key = self.client.topics.list_shared_access_keys(group.name, name)
+            self.primary_key = key.key1
+            self.endpoint = self.resource.endpoint
+        else:
+            self.resource = FakeResource(name=name, id=name)
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py:76]: I had a recollection there were missing properties from the fakeresource?  or was that mitigated in some other way.

File: sdk/eventgrid/azure-eventgrid/tests/test_eg_consumer.py [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+import pytest
+import json
+import datetime as dt
+
+from devtools_testutils import AzureMgmtTestCase
+from msrest.serialization import UTC
+from azure.eventgrid import EventGridConsumer, CloudEvent, EventGridEvent, StorageBlobCreatedEventData
+
+# storage cloud event
+cloud_storage_dict = {
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/tests/test_eg_consumer.py:19]: Nit: maybe put these in a _mocks file or something to that effect.

File: sdk/eventgrid/azure-eventgrid/CHANGELOG.md (hunk 5 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/CHANGELOG.md:5]: I'll do this in a later pr once things are settled - this one is sort of supposed to be a work in progress pr.

File: sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py (hunk 21 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py:76]: tldr it's mitigated in another way ;)

File: sdk/eventgrid/azure-eventgrid/swagger/README.PYTHON_T2.md (hunk 19 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/swagger/README.PYTHON_T2.md:1]: agree - this is misleading - changing it

File: sdk/eventgrid/azure-eventgrid/samples/consume_samples/e2e_samples/functionsapp/EventGridTrigger1/__init__.py (hunk 18 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/samples/consume_samples/e2e_samples/functionsapp/EventGridTrigger1/__init__.py:22]: not really - uncommenting

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_consumer.py (hunk 6 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_consumer.py:41]: can be a kwarg

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py (hunk 10 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:46]: there should be an rtype - will go through the docstrings and fix them in a later pr

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_shared/mixins.py (hunk 12 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_shared/mixins.py:7]: might be something that can move to azure core

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py (hunk 13 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:54]: should be a valueError

File: sdk/eventgrid/azure-eventgrid/tests/_mocks.py [hunk 23]
```diff
@@ -0,0 +1,74 @@
+# storage cloud event
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/tests/_mocks.py:1]:
```suggestion
import json

# storage cloud event
```

File: sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+                                                     playback_fake_resource=playback_fake_resource,
+                                                     client_kwargs=client_kwargs)
+        self.resource_group_parameter_name = resource_group_parameter_name
+        self.parameter_name = parameter_name
+        self.parameter_location = parameter_location
+        self.name_prefix = name_prefix
+        if random_name_enabled:
+            self.resource_moniker = self.name_prefix + "egtopic"
+        
+        self.set_cache(use_cache)
```
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py:50]:
```suggestion
        self.set_cache(use_cache, name_prefix)
```

File: sdk/eventgrid/azure-eventgrid/CHANGELOG.md (hunk 5 above)
Comment (KieranBrantnerMagee) [reply sdk/eventgrid/azure-eventgrid/CHANGELOG.md:5]: SGTM.  Let's make sure we have an issue tracking to put docs in before september release.  (Blame this on my distrust of my own memory to ensure this doesn't fall off the bus, obvious though it may be.)

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py (hunk 10 above)
Comment (KieranBrantnerMagee) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:46]: Echoing my normal mantra of "fine with that; let's just make sure we have an issue to track it."

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_models.py [hunk 25]
```diff
@@ -0,0 +1,199 @@
...
+    def model(self):
+        # type: () -> Union[CloudEvent, EventGridEvent]
+        """
+        Returns strongly typed EventGridEvent/CloudEvent object defined by the format of the properties.
+        All properties of the model are strongly typed (ie. for an EventGridEvent, event_time property will return a datetime.datetime object).
+
+        model.data: Returns a system event type(StorageBlobCreated, StorageBlobDeleted, etc.). If model.type/model.event_type is not defined in the 
+          system registry, returns None.
+
+        :rtype: Union[CloudEvent, EventGridEvent]
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_models.py:158]:
```suggestion
        :raise: :class:`ValueError`, when events do not follow CloudEvent or EventGridEvent schema.

        :rtype: Union[CloudEvent, EventGridEvent]
```

File: sdk/eventgrid/azure-eventgrid/CHANGELOG.md (hunk 5 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/CHANGELOG.md:5]: https://github.com/Azure/azure-sdk-for-python/issues/13223

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py (hunk 10 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:46]: https://github.com/Azure/azure-sdk-for-python/issues/13223

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+        self._topic_hostname = topic_hostname
+
+
+    async def send(self, events, **kwargs):
+        # type: (Union[List[CloudEvent], List[EventGridEvent], List[CustomEvent]], Any) -> None
+        """Sends event data to topic hostname specified during client initialization.
+
+        :param  events: A list of CloudEvent/EventGridEvent/CustomEvent to be sent.
+        :type events: Union[List[models.CloudEvent], List[models.EventGridEvent], List[models.CustomEvent]]
+        :rtype: None
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:43]: The send here and sync client need :raise: ValueError (I believe you had the more detailed exception message elsewhere too?)

File: sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py (hunk 20 above)
Comment (KieranBrantnerMagee) [reply sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py:50]: Still seems to be present?

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py (hunk 10 above)
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_publisher_client.py:46]:
```suggestion
        :rtype: None
        raise: :class:`ValueError`, when events do not follow specified SendType.
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py (hunk 26 above)
Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:43]:
```suggestion
        :rtype: None
        raise: :class:`ValueError`, when events do not follow specified SendType.
```

Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:43]: Added the raises for both sync and async - Trimmed the message to keep it a little more generic till we finalize the send spec (Pr open)

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/_consumer.py [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+_LOGGER = logging.getLogger(__name__)
+
+from ._models import DeserializedEvent
+
+class EventGridConsumer(object):
+    """
+    A consumer responsible for deserializing event handler messages, to allow for access to strongly typed Event objects.
+    """
+
+    def deserialize_event(self, event, **kwargs):
```
Comment (KieranBrantnerMagee) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/_consumer.py:30]: Given the direction conversations are going ( per discussions [here](https://gist.github.com/kerri-lee/900f7fcee8ba1bcca5ee18e19e9b59d3) ) it's looking like the shape of this API may be changing. Items such as:
- location, here or on the event models
  - With this, not utilizing a wrapper class (deserializedEvent) and instead doing the "lazy load" when accessing the data property.
- parsing directly from e.g. a requestContext

 We're going to need to have a longer conversation when @lmazuel is back.  I'm not too concerned if this is the base we work off of since this logic will presumably form the core regardless, especially since the discussion is still in flight, but was curious to hear your thoughts (@rakshith91), largely on whether my instinct to iterate from working to working since we're preview isn't disastrous, or if you'd wait to hit master until we locked fully.

Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/_consumer.py:30]: LEt this not block us since we haven't settled on design yet - have an issue tracking it
https://github.com/Azure/azure-sdk-for-python/issues/13227

File: sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py (hunk 20 above)
Comment (KieranBrantnerMagee) [reply sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py:50]: From our OOB conversation; there's buggy interplay between the cache and name_prefix.  I've taken an action item on myself to look into that, for now this is fine.
